internal/migration/migration.go

```diff
@@ -0,0 +1,478 @@
+	//nolint:gomnd // See below
+	case bufferSize > 10_000_000:
+		//nolint:lll // This line has the right amount of chars
+		logger.Warnw("buffer size seems a bit too large. Make sure your system can allocate enough system memory for this", "bufferSize", bufferSize)
+	}
+	return bufferSize, nil
+}
+
+func automaticBufferSize(logger log.Logger, sourceBeaconPath string) (int, error) {
+	logger.Warnw("buffer size set to 0. Running automatic buffer inference. Make sure you have enough system memory for this")
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return -1, err
+	}
+	defer func() {
+		_ = existingDB.Close()
+	}()
+
+	var bufferSize = 0
+
+	err = existingDB.View(func(tx *bbolt.Tx) error {
+		bufferSize = tx.Bucket(bucketName).Stats().KeyN
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	logger.Infow("buffer size automatically inferred from existing DB", "bufferSize", bufferSize)
+
+	return bufferSize, nil
+}
+
+//nolint:lll // This function has the right amount of
+func shouldMigrate(ctx context.Context, logger log.Logger, sourceBeaconPath, beaconName string, destination chain.StorageType, pgDSN string) error {
+	//nolint:exhaustive // We want to explicitly ignore the chain.MemDB backend since there's nothing to migrate there.
+	switch destination {
+	case chain.BoltDB:
+		return shouldMigrateBolt(sourceBeaconPath)
+	case chain.PostgreSQL:
+		return shouldMigratePostgres(ctx, logger, beaconName, pgDSN)
+	default:
+		return fmt.Errorf("unknown destination type %q for migration package", destination)
+	}
+}
+
+func shouldMigrateBolt(sourceBeaconPath string) error {
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = existingDB.Close()
+	}()
+
+	return existingDB.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(bucketName)
+		_, value := bucket.Cursor().First()
+		b := beacon{}
+		err := json.Unmarshal(value, &b)
+		if err != nil {
+			return ErrMigrationNotNeeded
+		}
+
+		return nil
+	})
+}
+
+func pgConn(ctx context.Context, logger log.Logger, beaconName, pgDSN string) (*pgdb.Store, func(), error) {
+	pgConf, err := database.ConfigFromDSN(pgDSN)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	conn, err := database.Open(ctx, pgConf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := schema.Migrate(ctx, conn); err != nil {
+		err := fmt.Errorf("migrating error: %w", err)
+		return nil, nil, err
+	}
+
+	store, err := pgdb.NewStore(ctx, logger, conn, beaconName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cancel := func() {
+		_ = store.Close(ctx)
+		_ = conn.Close()
+	}
+
+	return store, cancel, err
+}
+
+func shouldMigratePostgres(ctx context.Context, logger log.Logger, beaconName, pgDSN string) error {
+	store, cancel, err := pgConn(ctx, logger, beaconName, pgDSN)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	storeLen, err := store.Len(ctx)
```

Comment:
This is interesting, I'm guessing this is a query. Would have been nice if the api was `store.QueryLength(ctx)`

Reply:
Unfortunately, we can't change the API here. But yes, there's a query here.
This also applies to the BoltDB implementation.